src/lib/tls/asio/asio_stream.h

```diff
@@ -0,0 +1,585 @@
+       * @param buffers The buffers into which the data will be read.
+       * @return The number of bytes read. Returns 0 if an error occurred.
+       * @throws boost::system::system_error if error occured
+       */
+      template <typename MutableBufferSequence>
+      std::size_t read_some(const MutableBufferSequence& buffers)
+         {
+         boost::system::error_code ec;
+         auto const n = read_some(buffers, ec);
+         boost::asio::detail::throw_error(ec, "read_some");
+         return n;
+         }
+
+      /**
+       * Write some data to the stream. The function call will block until one or more bytes of data has been written
+       * successfully, or until an error occurs.
+       * @param buffers The data to be written.
+       * @param ec Set to indicate what error occurred, if any.
+       * @return The number of bytes written.
+       */
+      template <typename ConstBufferSequence>
+      std::size_t write_some(const ConstBufferSequence& buffers,
+                             boost::system::error_code& ec)
+         {
+         std::size_t sent = tls_encrypt_some(buffers, ec);
+         if(ec)
+            { return 0; }
+
+         writePendingTlsData(ec);
+         if(ec)
+            { return 0; }
+
+         return sent;
+         }
+
+      /**
+       * Write some data to the stream. The function call will block until one or more bytes of data has been written
+       * successfully, or until an error occurs.
+       *
+       * @param buffers The data to be written.
+       * @return The number of bytes written.
+       * @throws boost::system::system_error if error occured
+       */
+      template <typename ConstBufferSequence>
+      std::size_t write_some(const ConstBufferSequence& buffers)
+         {
+         boost::system::error_code ec;
+         auto const n = write_some(buffers, ec);
+         boost::asio::detail::throw_error(ec, "write_some");
+         return n;
+         }
+
+      /**
+       * Start an asynchronous write. The function call always returns immediately.
+       *
+       * @param buffers The data to be written.
+       * @param handler The handler to be called when the write operation completes. Copies will be made of the handler
+       *        as required. The equivalent function signature of the handler must be:
+       *        void(boost::system::error_code, std::size_t)
+       */
+      template <typename ConstBufferSequence, typename WriteHandler>
+      BOOST_ASIO_INITFN_RESULT_TYPE(WriteHandler,
+                                    void(boost::system::error_code, std::size_t))
+      async_write_some(const ConstBufferSequence& buffers, WriteHandler&& handler)
+         {
+         BOOST_ASIO_WRITE_HANDLER_CHECK(WriteHandler, handler) type_check;
+
+         boost::asio::async_completion<WriteHandler, void(boost::system::error_code, std::size_t)> init(handler);
+
+         boost::system::error_code ec;
+         std::size_t sent = tls_encrypt_some(buffers, ec);
```

Comment:
Does it make sense to move the encryption into the asynchronous operation as well? As far as I can see, `tls_encrypt_data` just manages the `buffers` and then calls `native_handle()->send()`. I think this should be doable in the `AsyncWriteOperation` as well. However, at the expense of duplicating the logic in `tls_encrypt_data` to a certain extend, no?

My main argument for this is symmetry though. The `AsyncReadOperation` takes care of the deciphering internally.